### PR TITLE
fix(transfers)!: Add txHash to transfer in statuses

### PR DESCRIPTION
**Pull request by jophish**
Per [this change](https://github.com/fiatconnect/specification/pull/95) to the specification, adds `txHash` as an optional field to transfer (in) statuses.

Since this PR encodes a good deal of semantics onto the transfer status endpoints/webhooks, this PR introduces a breaking change.

### CODE REVIEW

File: src/transfer.ts [hunk 1]
```diff
@@ -70,6 +70,7 @@ export const transferStatusResponseSchema = z.object(
     fiatAccountId: fiatAccountIdSchema,
     transferId: z.string(),
     transferAddress: z.string(),
+    txHash: z.string().optional(),
```
Comment (cajubelt) [inline src/transfer.ts:73]: you could make this required in the cases where it's required as follows:
- make the schema union of two schemas, maybe "preTxTransferStatusResponseSchema" and "postTxTransferStatusResponseSchema"
- restrict the transfer statuses allowable in each to the pre-tx statuses and post-tx statuses respectively
- make txHash required in postTx and omit it from preTx

**Review by cajubelt — COMMENTED** [review]
you could actually include the semantics you mention in the PR description

File: src/transfer.ts (hunk 1 above)
Comment (jophish) [reply src/transfer.ts:73]: Yeah, I had considered this, though there's somewhat of an inconsistency in the types repo re: enforcing these more nuanced semantics (e.g., we don't enforce that fiat account schemas in quote responses match the fiat account type the quote response claims they do), so I wasn't certain if it was worth enforcing this. On top of that, this approach has the quality of being non-breaking, which was incidentally kind of nice. 

That said, I'll update the schemas to enforce these semantics; it's not particularly difficult.

File: src/transfer.ts [hunk 2]
```diff
@@ -25,6 +25,45 @@ export const transferStatusSchema = z.nativeEnum(TransferStatus, {
   description: 'transferStatusSchema',
 })
 
+// These enums are first segmented between transfers in and transfers out
+// since they both allow a slightly different set of statuses.
+// Transfer in statuses are further segmented, since once a transfer in
+// progresses to sending the user funds, providers are expected
+// to return new metadata about the transfer. This segmentation allows
+// us to encode those semantics into Zod schemas.
+export enum TransferInStatusPreTx {
```
Comment (jophish) [inline src/transfer.ts:34]: I had originally wanted to "merge" native TS enums here; that is, create `TransferInStatusPreTx` and `TransferInStatusPostTx`, and have `TransferInStatus` be a new enum created by "merging" the first two. TLDR, this isn't possible in TS while playing nicely with Zod's `nativeEnum`, so I opted for a different approach here by re-using enum values so as not to duplicate string literals.

File: src/transfer.ts [hunk 3]
```diff
@@ -25,6 +25,45 @@ export const transferStatusSchema = z.nativeEnum(TransferStatus, {
...
+  TransferStarted = TransferStatus.TransferStarted,
+  TransferFiatFundsDebited = TransferStatus.TransferFiatFundsDebited,
+  TransferReceivedFiatFunds = TransferStatus.TransferReceivedFiatFunds,
+  TransferFailed = TransferStatus.TransferFailed,
```
Comment (jophish) [inline src/transfer.ts:38]: n.b. `TransferFailed` for transfers in _sometimes_ requires a TX hash to be present, but not always, so it's in the `TransferInStatusPreTx` enum (e.g., transfers can fail before crypto is sent, or after crypto is sent, if the blockchain transfer fails for some reason).

File: src/transfer.ts (hunk 1 above)
Comment (jophish) [reply src/transfer.ts]: ok, it wasn't particularly straightforward, either :sweat_smile:

File: src/transfer.ts [hunk 4]
```diff
@@ -25,6 +25,45 @@ export const transferStatusSchema = z.nativeEnum(TransferStatus, {
...
+}
+export const transferInStatusPreTxSchema = z.nativeEnum(TransferInStatusPreTx, {
```
Comment (cajubelt) [inline src/transfer.ts:40]:
```suggestion
export const transferInStatusPreTxSchema = z.enum(
  [
    TransferStatus.TransferStarted,
    TransferStatus.TransferFiatFundsDebited,
    TransferStatus.TransferReceivedFiatFunds,
    TransferStatus.TransferFailed,
  ],
```

Comment (cajubelt) [reply src/transfer.ts:40]: I found the `TransferInStatusPreTx` and native enums created in this PR kinda clunky, so I tried out a different approach. I think the downside is that if you actually wanted a typescript enum for TransferInStatusPreTx, you wouldn't have one, but I suspect the only actual use case for it is for typing, which you can get from the zod schema.

Comment (cajubelt) [reply src/transfer.ts:40]: feel free to adopt this approach or stick with what you've got

File: src/transfer.ts (hunk 2 above)
Comment (cajubelt) [reply src/transfer.ts:34]: i had a different idea for this that may interest you, commented on it below

**Review by cajubelt — APPROVED** [review]

**Pull request merged**